annif/backend/nn_ensemble.py

```diff
@@ -0,0 +1,126 @@
+"""Neural network based ensemble backend that combines results from multiple
+projects."""
+
+
+import os.path
+import numpy as np
+from tensorflow.keras.layers import Input, Dense, Add, Flatten, Lambda, Dropout
+from tensorflow.keras.models import Model, load_model
+import tensorflow.keras.backend as K
+import annif.corpus
+import annif.project
+import annif.util
+from annif.exception import NotInitializedException
+from annif.suggestion import VectorSuggestionResult
+from . import ensemble
+
+
+class NNEnsembleBackend(ensemble.EnsembleBackend):
+    """Neural network ensemble backend that combines results from multiple
+    projects"""
+
+    name = "nn_ensemble"
+
+    MODEL_FILE = "nn-model.h5"
+
+    DEFAULT_PARAMS = {
+        'nodes': 60,
+        'dropout_rate': 0.2,
+        'optimizer': 'adam',
+        'epochs': 10,
+    }
+
+    # defaults for uninitialized instances
+    _model = None
+
+    def default_params(self):
+        params = {}
+        params.update(super().default_params())
+        params.update(self.DEFAULT_PARAMS)
+        return params
+
+    def initialize(self):
+        if self._model is not None:
+            return  # already initialized
+        model_filename = os.path.join(self.datadir, self.MODEL_FILE)
+        if not os.path.exists(model_filename):
+            raise NotInitializedException(
+                'model file {} not found'.format(model_filename),
+                backend_id=self.backend_id)
+        self.debug('loading Keras model from {}'.format(model_filename))
+        self._model = load_model(model_filename)
+
+    def _merge_hits_from_sources(self, hits_from_sources, project, params):
+        score_vector = np.array([hits.vector * weight
+                                 for hits, weight in hits_from_sources])
+        results = self._model.predict(
+            np.expand_dims(score_vector.transpose(), 0))
+        return VectorSuggestionResult(results[0], project.subjects)
+
+    def _create_model(self, sources, project):
```

Comment:
Any comments @juhoinkinen or @mvsjober to the Keras model defined here? Would you do something differently?

Reply:
Looks good